src/controllers/bid.py

```diff
@@ -1,27 +1,131 @@
-from src.actions.game_actions import bid_on_power_plant, next_player
+from src.actions.game_actions import pass_current_bid, pass_initial_bid, set_current_bid, \
+    set_current_player, win_current_bid, next_phase, clear_bid, clear_bought_or_passed
+from src.actions.players_actions import remove_power_plant
+from src.decorators.connect import connect
 
 
-def bid(data, store):
+@connect
+def bid(bid_request, **kwargs):
     """ Handle how to update the game when a player tried to bid on a market item"""
-    player_id = data.get('player_id')
-    amount = data.get('amount')
-    card = data.get('card')
+    player_id = bid_request.get('player_id')
+    amount = bid_request.get('amount')
+    card = bid_request.get('card')
 
-    state = store['get_state']()
-    current_bid = state.get('game').get('current_bid')
+    get_state = kwargs['get_state']
+    dispatch = kwargs['dispatch']
+
+    initial_state = get_state()
+    if initial_state.get('game').get('phase') != 2:
+        return False
+    num_players = len(initial_state.get('players'))
+    current_bid = initial_state.get('game').get('current_bid')
+    # This indicates a passing bid
+    if amount == 0:
+        # If no one has bid, this indicates passing on an initial bid
+        if not current_bid:
+            dispatch(pass_initial_bid(player_id))
+        else:
+            dispatch(pass_current_bid(player_id))
+    else:
+        if not valid_bid(player_id, amount, card, initial_state.get('game'), initial_state.get('players'), initial_state.get('market')):
+            return False
+
+        dispatch(bid(player_id, card, amount))
+
+    if bid_is_won(get_state().get('game'), num_players):
+        dispatch(win_current_bid())
+        dispatch(clear_bid())
+
+    if get_state().get('game').get('bought_or_passed') == num_players:
+        dispatch(next_phase())
+        dispatch(clear_bought_or_passed())
+        dispatch(clear_bid())
+
+    dispatch(set_current_player(get_next_player(get_state().get('game'), get_state().get('players'))))
+
+    return True
+
+
+@connect
+def discard_power_plant(discard_request, **kwargs):
+    """ Handle discarding a powerplant when a player has 4"""
+    player_id = discard_request.get('player_id')
+    card = discard_request.get('card')
+
+    get_state = kwargs['get_state']
+    dispatch = kwargs['dispatch']
+
+    initial_state = get_state()
+    if not card:
+        return False
+
+    player = [pl for pl in initial_state.get('game').get('players') if pl.get('player_id') == player_id][0]
```

Comment:
This could blow up if the player_id supplied is wrong, because you'd be trying to get the zeroth element of an empty list.

I'd just add one more check:
```
if player_id not in initial_state.get('player_rank'): 
  return False
```
Because we'll always have a list of the player_ids in player rank.  If that check passes, then you know you'll get the player.

It also might be useful to abstract getting the player from the list of players into a helper method, as I'm sure we'll be doing that a lot